MoreLinq/DeconstructibleEnumerable.cs

```diff
@@ -0,0 +1,86 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2019 Atif Aziz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+namespace MoreLinq
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+
+    /// <summary>
+    /// Represents a source whose elements can be sequentially deconstructed
+    /// into separate variables.
+    /// </summary>
+    /// <typeparam name="T">The type of the elements.</typeparam>
+
+    public partial struct DeconstructibleEnumerable<T> : IEnumerable<T>
+    {
+        readonly IEnumerable<T> _source;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="DeconstructibleEnumerable{T}"/>
+        /// with a sequence.
+        /// </summary>
+        /// <param name="source">
+        /// The source sequence on which to enable deconstruction.</param>
+
+        public DeconstructibleEnumerable(IEnumerable<T> source) =>
+            _source = source ?? throw new ArgumentNullException(nameof(source));
+
+        /// <inheritdoc />
+        public IEnumerator<T> GetEnumerator() =>
+            _source != null ? _source.GetEnumerator()
+                            : throw CreateTooShortError();
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        void Deconstruct(int count, out T item1, out T item2, out T item3, out T item4, out T item5, out T item6, out T item7, out T item8, out T item9, out T item10, out T item11, out T item12, out T item13, out T item14, out T item15, out T item16)
```

Comment:
I can't find where this method is used, it appears it isn't.

Reply:
This was leftover from an initial spike that was modelled after [`FoldImpl`](https://github.com/morelinq/MoreLINQ/blob/66b6bc4525201da0d763b51c4b91dcf4315df9d0/MoreLinq/Fold.cs#L25-L90) and thanks for spotting it. Removed in ad0df41c6852a1e0f88157d2a0456ef7f1d7b7c4.